modulus/models/afno/modafno.py

```diff
@@ -0,0 +1,649 @@
+            latent_features=latent_features,
+            out_features=out_features,
+            activation_fn=activation_fn,
+            drop=drop,
+        )
+        if scale_shift_kwargs is None:
+            scale_shift_kwargs = {}
+        self.scale_shift = ScaleShiftMlp(
+            mod_features, latent_features, **scale_shift_kwargs
+        )
+
+    def forward(self, x: Tensor, mod_embed: Tensor) -> Tensor:
+        (scale, shift) = self.scale_shift(mod_embed)
+
+        scale_shift_shape = (scale.shape[0],) + (1,) * (x.ndim - 2) + (scale.shape[1],)
+        scale = scale.view(*scale_shift_shape)
+        shift = shift.view(*scale_shift_shape)
+
+        x = self.fc1(x)
+        x = x * scale + shift
+        x = self.act(x)
+        x = self.drop(x)
+        x = self.fc2(x)
+        x = self.drop(x)
+        return x
+
+
+class ModAFNO2DLayer(AFNO2DLayer):
+    """AFNO spectral convolution layer
+
+    Parameters
+    ----------
+    hidden_size : int
+        Feature dimensionality
+    mod_features : int
+        Number of modulation features
+    num_blocks : int, optional
+        Number of blocks used in the block diagonal weight matrix, by default 8
+    sparsity_threshold : float, optional
+        Sparsity threshold (softshrink) of spectral features, by default 0.01
+    hard_thresholding_fraction : float, optional
+        Threshold for limiting number of modes used [0,1], by default 1
+    hidden_size_factor : int, optional
+        Factor to increase spectral features by after weight multiplication, by default 1
+    scale_shift_kwargs : dict, optional
+        Options to the MLP that computes the scale-shift parameters
+    scale_shift_mode: str
+        If 'complex' (default), compute the scale-shift operation using complex
+        operations. If 'real', use real operations.
+    """
+
+    def __init__(
+        self,
+        hidden_size: int,
+        mod_features: int,
+        num_blocks: int = 8,
+        sparsity_threshold: float = 0.01,
+        hard_thresholding_fraction: float = 1,
+        hidden_size_factor: int = 1,
+        scale_shift_kwargs: Union[dict, None] = None,
+        scale_shift_mode: str = "complex",
+    ):
+        super().__init__(
+            hidden_size=hidden_size,
+            num_blocks=num_blocks,
+            sparsity_threshold=sparsity_threshold,
+            hard_thresholding_fraction=hard_thresholding_fraction,
+            hidden_size_factor=hidden_size_factor,
+        )
+
+        self.scale_shift_mode = scale_shift_mode
+        self.channel_mul = 1 if scale_shift_mode == "real" else 2
+        if scale_shift_kwargs is None:
+            scale_shift_kwargs = {}
+        self.scale_shift = ScaleShiftMlp(
+            mod_features,
+            self.num_blocks
+            * self.block_size
+            * self.hidden_size_factor
+            * self.channel_mul,
+            **scale_shift_kwargs,
+        )
+
+    def forward(self, x: Tensor, mod_embed: Tensor) -> Tensor:
```

Comment:
May consider adding an assert check here to check consistency between x and mod_embed

Reply:
This would be an assert for the tensor channel dimension? In any case, I tried to add it and ruff complains:
```modulus/models/afno/modafno.py:191:9: S101 Use of `assert` detected```

Reply:
Ah yeah I forgot about the asset checks, ruff will want you to have a if statement and then raise the error yourself.

I suppose the idea is to make sure that the user's x and mod_embed are of the correct sizes and give a easy to understand error if it isnt. Could help make it easier to use for people. Just my opinion though,Thoughts?